### DISCUSSION TITLE
Line Art questions

### DISCUSSION THREAD

**Discussion by aVersionOfReality** [bnpr/Malt#126] [Q&A]
Hello, a few questions on using line art.

1) Why are the lines not fully black when at thinner sizes? The lines don't display as fully black until about size 4, which is quite thick depending on zoom level.
![image](https://user-images.githubusercontent.com/36803956/133580299-8d74ee1e-6100-464b-9123-2699295d56dd.png)

Here is another example where I mapped Z depth to match the character and used it to change the thickness. The narrower lines further away become less dark.
![image](https://user-images.githubusercontent.com/36803956/133580979-ce5a96fc-b3b6-4f9a-ae20-0f45d64b9d2f.png)

2) Is there any existing way to setup Relative line thickness, such as we have in Cycles ToonKit? If it can't currently be done, what is missing to be able to?

3) It is possible to get dotted or broken lines, especially with normal/angle based lines, where it doesn't consistently pick up an edge. Is there a way to prevent this, or to remove it? Perhaps a way to check neighboring pixels? This picture is set to 1 sample to make them easier to see. With multiple samples, they blur out a bit, but don't fully disappear. It especially causes issues if beveling the normals, as this effectively blurs these fragments and makes them even more visible.
![image](https://user-images.githubusercontent.com/36803956/133581574-cfdc90b5-b9db-49d0-ad2c-987ef9421b95.png)

4) What exactly is the difference between plugging my lineart node's output into line_width on the PixelOutput node and just using the lines as color, like this? (the mix is just to invert the color, otherwise the lines would be white.)
![image](https://user-images.githubusercontent.com/36803956/133695839-eeaa569e-bb59-482f-b5b7-313d7e697be3.png)

Thanks!

**Comment by pragma37** [comment]
1. There are 2 reasons for this.
- Since line detection is pixel-based, I sacrificed a bit of width precision to make sure that subpixel width variations and varying width lines are smooth. There may be better ways to handle this, but that's what I came up with.
- The multiple samples are accumulated in linear space, so it's easy for white to eat the blacks.

As a workaround, you can use darker-than-black line colors to compensate.

![imagen](https://user-images.githubusercontent.com/36610999/133858040-42f5f1a3-c96f-4ed7-9b0a-d43b24b31c13.png)

For scaling your line based on distance, you might get better results at different FoVs and resolutions using pixel_world_size.

2. Never used ToonKit. How does that work?

3. You may get better results leaving some room between line detection angle thresholds and angle widths, so it's not a purely boolean operation.

4. Lines are expanded in a post-process pass. So if you just apply them directly to the pixel color the lines will always be 1 pixel wide.

**Reply by aVersionOfReality** [reply]
2: The lines stay the same thickness relative to the geometry regardless of camera. I'm not sure how the math actually works. ToonKit is all OSL, so I assume we can look in the script file and find out, license allowing. I will look into it.

3: I'm not sure what you mean by thresholds and widths. The ones on the get_line_advanced node? They do not seem to help with this issue. It occurs are certain normal thresholds, even with get_line_simple.
![image](https://user-images.githubusercontent.com/36803956/133879420-ca126c2c-342d-4fbe-99ee-bfcf14d02c64.png)

4: Ah, right. So that means I cannot use lines to, say, create masks for parameters in a material if I want more thickness? I know that it is necessary to expand them in post for them to overlap off the model. Is that the only reason its done there? Could there be a version of lines that are expanded within the material?

And speaking of, where can Screen Shaders actually be used? I see I can make the node trees of that type, but have not been able to find where I actually assign them. (EDIT: Nvm, found it from your sample file. So its in the Render Layer nodes...)

Thanks!

**Reply by pragma37** [reply]
2. I meant what the option does, not how it's implemented. If you want to specify the line in world space units, just divide the line width by pixel_world_size.

3. get_line_simple is effectively the same as get_line_advanced where the min and max widths are the same. I just checked your file and I think it may be related to the non-manifold geometry of the hair.

4. It's not just for expanding the line outside of the model, it's for expanding the line at all. You can't just write to one pixel from another one. Any effect that requires reading from other pixels has to be done in multiple passes, so you can read the results of the previous pass from the next one.

**Comment by JPLost** [comment]
Is sketch-like lines achievable with get_line simple or advanced?
Can I offset the lines and create sketching like effect?
Or are these set on post-process pass? What is defined as post-process pass?

As far as I can see, the node only allows even strokes to be used. To add variation the code would need to be expanded as shown in freestyle strokes http://freestyle.sourceforge.net/doc/html/Stroke_8h-source.html

Lets also add this here.
https://www.forceflow.be/dl/thesis_kul/masterproef.pdf

**Reply by pragma37** [reply]
*Edit:* Up to date example (https://github.com/bnpr/Malt/discussions/305#discussioncomment-2834732)

Do you mean offsetting the width? You can just scale the line width however you want:

![imagen](https://user-images.githubusercontent.com/36610999/134017278-1ebb78da-b720-4f6f-9d44-224ce01060c2.png)

BTW, for the next time, it's better if you just open a new question.
